apm-agent-plugins/apm-redis-plugin/apm-lettuce-plugin/src/main/java/co/elastic/apm/agent/redis/lettuce/LettuceStopSpanInstrumentation.java

```diff
@@ -0,0 +1,145 @@
+        return nameStartsWith("io.lettuce.core.")
+            .and(hasSuperType(named("io.lettuce.core.protocol.RedisCommand")));
+    }
+
+    @Override
+    public Collection<String> getInstrumentationGroupNames() {
+        return Arrays.asList("redis", "lettuce");
+    }
+
+    public static class OnComplete extends LettuceStopSpanInstrumentation {
+
+        @Advice.OnMethodEnter(suppress = Throwable.class)
+        private static void beforeComplete(@Advice.Local("nested") boolean nested) {
+            nested = LettuceUtil.beforeComplete(null, false);
+        }
+
+        @Advice.OnMethodExit(suppress = Throwable.class)
+        private static void afterComplete(@Advice.Local("nested") boolean nested) {
+            LettuceUtil.afterComplete(nested);
+        }
+
+        @Override
+        public ElementMatcher<? super MethodDescription> getMethodMatcher() {
+            return named("complete").and(takesArguments(0));
+        }
+    }
+
+    public static class OnCompleteExceptionally extends LettuceStopSpanInstrumentation {
+
+        @Advice.OnMethodEnter(suppress = Throwable.class)
+        private static void beforeComplete(@Advice.Local("nested") boolean nested,
+                                           @Advice.Argument(0) Throwable throwable) {
+            nested = LettuceUtil.beforeComplete(throwable, false);
+        }
+
+        @Advice.OnMethodExit(suppress = Throwable.class)
+        private static void afterComplete(@Advice.Local("nested") boolean nested) {
+            LettuceUtil.afterComplete(nested);
+        }
+
+        @Override
+        public ElementMatcher<? super MethodDescription> getMethodMatcher() {
+            return named("completeExceptionally").and(takesArguments(Throwable.class));
+        }
+    }
+
+    public static class OnCancel extends LettuceStopSpanInstrumentation {
+
+        @Advice.OnMethodEnter(suppress = Throwable.class)
+        private static void beforeComplete(@Advice.Local("nested") boolean nested) {
+            nested = LettuceUtil.beforeComplete(null, true);
+        }
+
+        @Advice.OnMethodExit(suppress = Throwable.class)
+        private static void afterComplete(@Advice.Local("nested") boolean nested) {
+            LettuceUtil.afterComplete(nested);
+        }
+
+        @Override
+        public ElementMatcher<? super MethodDescription> getMethodMatcher() {
+            return named("cancel").and(takesArguments(0));
+        }
+    }
+
+    public static class LettuceUtil {
+
+        private static ThreadLocal<Boolean> nestedThreadLocal = new ThreadLocal<>();
+
+        @VisibleForAdvice
+        public static boolean beforeComplete(@Nullable Throwable t, boolean cancelled) {
+            try {
+                // guard against wrapped RedisCommands
+                if (nestedThreadLocal.get() == Boolean.TRUE || tracer == null) {
+                    return true;
+                }
+                TraceContextHolder<?> active = tracer.getActive();
```

Comment:
So, the command methods are guaranteed to be called within the `RedisChannelWriter#write` execution? If so, can't the `RedisChannelWriter#exit` be used to end the span, without worrying about nesting (still requires the instrumentation of `completeExceptionally` for the Exception probably)? If not- relying here on `getActive` is unsafe and may cause span leak.